Enable 8byte vector loading for fp16/bf16

jeetkanjani7 opened the pull request:
Test Plan:
Tested via local benchmarks and e2e runs. The bandwidth improves by 2.5x on A100 80GB gpu for 2 byte data types (fp16, bf16). Also significant improvement in e2e qps.
 {F1977455759}

Differential Revision: D73225699

Skylion007 commented on aten/src/ATen/native/cuda/Indexing.cu:
```diff
@@ -1706,7 +1750,13 @@ void index_select_out_cuda_impl(
         } else {
-          LARGE_INDEX(scalar_t, index_t, unsigned int, -1, -1, -1, true);
+
+          if (indexIsMajor) {
```
```suggestion
          if constexpr (indexIsMajor) {
```

Skylion007 commented on aten/src/ATen/native/cuda/Indexing.cu:
```diff
@@ -1563,11 +1568,60 @@ __global__ void indexSelectLargeIndex(cuda::detail::TensorInfo<T, IndexType> dst
+        if (li >= totalSize) break;
+
+        IndexType dstIndex2, elem2;
+        if (IndexIsMajor) {
```
```suggestion
        if constexpr (IndexIsMajor) {

jeetkanjani7 committed:
Enable 8byte vector loading for fp16/bf16 (#152242)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/152242

Test Plan:
Tested via local benchmarks and e2e runs. The bandwidth improves by 2.5x on A100 80GB gpu for 2 byte data types (fp16, bf16). Also significant improvement in e2e qps.
 {F1977455759}

Differential Revision: D73225699

eqy commented on aten/src/ATen/native/cuda/Indexing.cu:
```diff
@@ -1539,9 +1539,14 @@ __global__ void indexSelectLargeIndex(cuda::detail::TensorInfo<T, IndexType> dst
+  constexpr bool kPack4 = (sizeof(T) == 2);
+  constexpr int kElemsPerVec = kPack4 ? 4 : 1;
+
+  for (IndexType vecLinear = (blockIdx.x * blockDim.x + threadIdx.x) * kElemsPerVec;
```
Is there any risk of overflow (can `IndexType` be 64-bit wide here)? If so, then `blockIdx.x * blockdim.x + threadIdx.x` should probably be cast to `IndexType` as they are implicitly 32-bit here.
